### ISSUE TITLE
ImportError...circular import

### ISSUE THREAD

**Issue by wwoofbum** [miguelgrinberg/microblog#367]
Kindle version, 2024, chapter 6:

errors.py:

```
from flask import render_template
from app import app, db


@app.errorhandler(404)
def not_found_error(error):
    return render_template('404.html'), 404


@app.errorhandler(500)
def internal_error(error):
    db.session.rollback()
    return render_template('500.html'), 500
```
```
__init__.py:

from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_migrate import Migrate
from flask_login import LoginManager
from config import Config

app = Flask(__name__)
app.config.from_object(Config)
db = SQLAlchemy(app)
migrate = Migrate(app, db)
login = LoginManager(app)
login.login_view = 'login'

from app import routes, models, errors
```

results in error:
...
File 'home/owner/microblog/app/__init__.py' line 14 in module
  from app import routes, models, errors
ImportError: cannot import name 'errors' from partially initialized module 'app' (most likely due to a circular import)
,,,

**Comment by miguelgrinberg** [comment]
The working code is available here on this repository. Have you tried it?

To understand a circular import the entire stack trace is needed. A single frame does not provide enough information.

**Comment by wwoofbum** [comment]
Well, my apologies. After continuing to develop the app, and attempting to go back to the state in which I encountered the problem, I am unable to replicate the error.
I had completed the development through chapter 6, and the app was performing as expected.  I began implementing the changes as I worked through chapter 7, and was at the point where errors.py, 404.html, and 500.html were in place, but before the sections in which emailing and logging were implemented.
Again, my apologies.  I appreciate the detail to which you have gone in your tutorial.
Gordon
    On Friday, March 1, 2024 at 01:02:40 PM PST, Miguel Grinberg ***@***.***> wrote:  
 
 


The working code is available here on this repository. Have you tried it?

To understand a circular import the entire stack trace is needed. A single frame does not provide enough information.

**Comment by wwoofbum** [comment]
Right.

A side note: Are you interested in copyedit notes regarding the Kindle (2024) edition of the MegaTutorial?  e.g.: First paragragh under heading 'Filters' of chapter 8, 'alot more data that I really want.' Should read 'alot more data than I really want.'
If uninterested, simply ignore this message.
    On Saturday, March 2, 2024 at 09:36:32 AM PST, Miguel Grinberg ***@***.***> wrote:  
 
 


Closed #367 as completed.
